bundles/org.openhab.binding.smaenergymeter/src/main/java/org/openhab/binding/smaenergymeter/internal/packet/DefaultPacketListenerRegistry.java

```diff
@@ -0,0 +1,85 @@
+/**
+ * Copyright (c) 2010-2024 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.smaenergymeter.internal.packet;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.openhab.binding.smaenergymeter.internal.packet.PacketListener.ReceivingTask;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of packet listener registry which manage multicast sockets.
+ *
+ * @author Łukasz Dywicki - Initial contribution
+ */
+@Component
+public class DefaultPacketListenerRegistry implements PacketListenerRegistry {
+
+    private final Logger logger = LoggerFactory.getLogger(DefaultPacketListenerRegistry.class);
+    private final ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1,
+            (runnable) -> new Thread(runnable, "sma-energy-meter-receiver"));
+    private final Map<String, PacketListener> listeners = new ConcurrentHashMap<>();
+
+    @Override
+    public PacketListener getListener(String group, int port) throws IOException {
+        String identifier = group + ":" + port;
+        PacketListener listener = listeners.get(identifier);
+        if (listener == null) {
+            listener = new PacketListener(this, group, port);
+            listeners.put(identifier, listener);
+        }
+        return listener;
+    }
+
+    @Deactivate
+    protected void shutdown() throws IOException {
+        for (Entry<String, PacketListener> entry : listeners.entrySet()) {
+            try {
+                entry.getValue().close();
+            } catch (IOException e) {
+                logger.warn("Multicast socket {} failed to terminate", entry.getKey(), e);
+            }
+        }
+        scheduler.shutdownNow();
+    }
+
+    public ScheduledFuture<?> addTask(Runnable runnable, int intervalSec) {
+        return scheduler.scheduleAtFixedRate(runnable, 0, intervalSec, TimeUnit.SECONDS);
```

Comment:
https://www.openhab.org/docs/developer/guidelines.html#e-runtime-behavior

scheduleWithFixedDelay is preferred to scheduleAtFixedRate. is scheduleAtFixedRate absolutely required here ?

Reply:
I guess `scheduleWithFixedDelay` is good enough, maybe @splatch can comment if the far more expensive `scheduleAtFixedRate` is essential.

Reply:
Changed.

Reply:
Sorry, I am stuck in preparations to industrial trade fair which takes place next week. Meter side push data every 100-200 ms (AFAIR), so ideally we should really pick them and publish at fixed rate. The fixed delay proposed in official docs is wrong in many places, cause it introduces random delays which are out-of-control for anyone. 

In context of this particular change it is irrelevant, cause it is not for receiving data, but publishing which is effortless and takes little to almost no-time.

Reply:
PR was merged with `scheduleWithFixedDelay` before I saw your message.

If this is really important, a new PR can be submitted.

Reply:
No problem, its fine to remain as-is.

Reply:
Thanks for assuring this is ok!